hyperdome/server/web.py

```diff
@@ -235,61 +220,132 @@ def counselor_signup():
+                "messages": [],
+                "sequence": 0,
+            }
+
+        @app.route("/hyperdome/api/v1/chats/<string:user_id>", methods=["GET"])
+        def get_key_data(user_id):
+            try:
+                if user_id in self.counselors:
+                    partner_id = self.counselors[user_id]["connected_guest"]
+                elif user_id in self.guests:
+                    partner_id = self.guests[user_id]["connected_counselor"]
+                else:
+                    abort(404, message="no user found for user id")
+                    return
+            except KeyError:
+                abort(404, message="user has no chat partner")
+                return
+            try:
+                chat = self.chats[partner_id]
+                key_data = chat["key_data"]
+                signature = chat["signature"]
+            except:
+                abort(404, message="partner has not yet initialized chat")
+                return
             else:
-                return "user left", 404
-            return "Success"
+                return jsonify(key_data=key_data, signature=signature)
 
-        @app.route("/collect_messages", methods=["GET"])
-        def collect_messages():
-            user_id = request.form["user_id"]
-            messages = self.pending_messages.pop(user_id, "")
+        @app.route(
+            "/hyperdome/api/v1/chats/<string:user_id>/<int:message_no>/",
+            methods=["POST"],
+        )
+        def message_from_user(user_id, message_no):
+            messages: list = request.json["messages"]
             try:
-                chat_status = (
-                    "CHAT_OVER"
-                    if self.active_chat_user_map[user_id] == ""
-                    else "CHAT_ACTIVE"
-                )
+                chat_partner = self.chats[user_id]["partner"]
+                self.chats[chat_partner]["messages"] += messages
+                self.chats[chat_partner]["sequence"] = message_no
             except KeyError:
-                chat_status = "NO_CHAT"
-            return jsonify(chat_status=chat_status, messages=messages)
+                abort(404, message="couldn't find message destination")
+            return jsonify(message="Success", no_messages=len(messages))
 
-    def error404(self):
-        self.add_request(Web.REQUEST_OTHER, request.path)
-        if request.path != "/favicon.ico":
-            self.error404_count += 1
+        @app.route(
+            "/hyperdome/api/v1/chats/<string:user_id>/<int:message_no>/",
+            methods=["GET"],
+        )
+        def collect_messages(user_id, message_no):
+            try:
+                messages = self.chats[user_id]["messages"][message_no:]
+                no_messages = len(messages)
+                next_message = message_no + no_messages
+            except KeyError:
+                abort(404, message="no chat")
+                return  # should be inaccessible
```

Comment:
My preference is to just put an `assert False, "Shouldn't get here"`
Up to you tho

Reply:
I was putting it there because my linter wasn't recognizing that 'abort' was going to leave the function, and was complaining that the variables I declared above may be undeclared. I'll check if `assert False` gets recognized as effectively exiting, otherwise I might reorganize the try blocks, maybe use the else, I just don't want a pyramid of doom, and I think functions that return in the middle, with a bunch of error handling after the intended return path, are much harder to understand.

Reply:
hey the linter does know that escapes code, and since it's assert it gets totally optimized out during builds, good call.